keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,362 @@
+Since we are migrating in-tree volume plugins out of tree, the health checking methods should be supported in CSI.
+
+The whole architecture is as below:
+![pv health monitor architecture](./pv-health-monitor.png)
+
+
+## Implementation
+
+### API change
+
+We plan to introduce three taints for PV health condition.
+
+**One is `VolumeNotAttached` taint**,
+```
+Key: VolumeNotAttached 
+Value: true
+VolumeTaintEffect: NoEffect  (Make VolumeTaintEffect non-required field)
+```
+
+If the volume is not attached now, we need to taint the PV using the VolumeNotAttached taint. 
+
+There is already a self-healing feature in-tree that checks if a volume is still attached.  It checks every minute.  If not attached, it tries to attach again. 
+
+If the self-healing controller reattaches successfully, it should remove the VolumeNotAttached taint then. 
+
+In order not to affect the self healing feature, we set the VolumeTaintEffect to NoEffect.
+
+**Another one is `VolumeNotMounted` taint**,
+```
+Key: VolumeNotMounted 
+Value: true
+VolumeTaintEffect: NoEffect  (Make VolumeTaintEffect non-required field)
+```
+
+This can be used for both device mount and volume mount checking. If the volume is not mounted, we need to taint it.
+
+For in use volumes, if the volumes are not mounted, reactions should be taken. 
+
+The reactions needs further discussion, and is not in the scope of this doc. 
+
+For unused volumes, the volume manager will try to mount again when we want to use it, so we do not need any effect.
+
+In summary, we set the VolumeTaintEffect to NoEffect too.
+
+**The last one is: `VolumeNotUsable` taint**,
+```
+Key: VolumeNotUsable (e.g. bad blocks, filesystem corrupted, etc)
+Value: true
+VolumeTaintEffect: NoAttachOrMount
+```
+
+If the volume is not usable because of bad blocks, filesystem corruption, disk deletion or any other reasons, we need to taint the PV using this taint.
+
+Here, we use NoAttachOrMount effect mainly for multi-attach volumes. 
+
+Because NoAttachOrMount effect will prevent new pods from using the tainted PV. 
+
+And as mentioned above, for in use volumes, reactions are also needed.
+
+When we finish discussing the reaction for unhealthy PVs, we need to revisit all the VolumeTaintEffects.
+
+### Networked storage
+
+For networked storage, the following volume health conditions will be checked:
+* Checks if volume still exists and/or is attached
+* Checks if volume is still mounted and usable
+
+Container Storage Interface (CSI) specification will be modified to add two RPCs ControllerCheckVolume and NodeCheckVolume. A monitor controller (external-monitor) that is responsible for watching the PersistentVolumeClaim, PersistentVolume, and VolumeAttachment API objects will be added. The external-monitor can be deployed as a sidecar container together with the CSI driver.  For PVs, the external-monitor calls CSI driver’s monitor interface to check volume health status.
+
+* ControllerCheckVolume RPC
```

Comment:
This seems to define a poll model. Would it also be possible for a push mechanism? I would assume that many storage systems underlying the CSI would actively notice and react to issues, and be able to notify upper layers in an expedient way. Thus, it should be possible for the CSI to communicate this. This is similar to the SCSI device specs, where a physical drive or physical SCSI controller can raise a notification when it detects an issue, so that upper layers can take immediate action. Being expedient can be important for applications that depend on application level replication and needing to know when to kick in rebuilds to reduce downtime.

Reply:
That's a good point.  Let me think about this.